### ISSUE TITLE
Why there are two separate embeddings from and to target vocabluary?

### ISSUE THREAD

**Issue by okuchaiev** [google/seq2seq#148]
Currently, seq2seq with attention model has 3 separate embedding layers:
1) model/att_seq2seq/encode/source_embedding/W  (37007x512, 18.95m/18.95m params) - to map source language word to their representations 
2) model/att_seq2seq/decode/target_embedding/W (512x37007, 18.95m/18.95m params) - to map target language words to their representations
3) model/att_seq2seq/decode/attention_decoder/decoder/logits/weights (37007x512, 18.95m/18.95m params) - to effectively map representation to target language word while doing prediction. This is fed into softmax together with bias.

Note that while all 3 are optimized, (2) isn't even used during inference.

Can't we use constrain (3) to just be transpose of (2) tf.transpose(model/att_seq2seq/decode/target_embedding/W) ? That is, learn only one embedding matrix per source and target language. @dennybritz could you please comment on this model design choice?

**Comment by dennybritz** [comment]
I believe this is standard.

2. Is used during inference. You always need to embed the word predicted by the decoder.
3. I don't look at this as an "embedding", more as just a weight matrix that is used to calculate logits, scores for each word. It doesn't necessarily contain meaningful representations of words. Also, the fact that this is the same size as 1/2 is just arbitrary because all your layers use 512 units. It doesn't need to be the same size, and generally isn't.

> Can't we use constrain (3) to just be transpose of (2) 

I guess you could, but to me it seems like that would unnecessarily constrain the model. I haven't really seen that being done, but could be an interesting experiment.

**Comment by okuchaiev** [comment]
> You always need to embed the word predicted by the decoder.
- Right, I missed that.

The fact that sizes of (1) and (2) match is just a coincidence. But it isn't obvious to me that sizes of (2) and (3) does not need to match. It seems to me that for translation they do need to match because they essentially act as one-way bridges between target language and decoder hidden states. Maybe they aren't necessarily the same for some other task?

Of course, if we constrain (3)=transpose(2) this will result in less free parameters. But such constraint looks natural to me. Anyway, hope I'll have time to try :)

**Comment by dennybritz** [comment]
>  But it isn't obvious to me that sizes of (2) and (3) does not need to match. It seems to me that for translation they do need to match because they essentially act as one-way bridges between target language and decoder hidden states. Maybe they aren't necessarily the same for some other task?

No, it's completely arbitrary. For example, you could just add another fully connected layer of size X after the last decoder state, then your matrix will be [X, vocab_size]. It will work just as well. Also, you could have different RNN layers be different sizes.

**Comment by okuchaiev** [comment]
> For example, you could just add another fully connected layer of size X after the last decoder state, then your matrix will be [X, vocab_size]

Then those "two" fully connected layers look like bridge to me from decoder hidden state to target vocab via softmax.  And if different RNN layers in decoder have different sizes then, yes, it will break the symmetry. 

My main point is that for decoder, there seems to be a need for a bridge from target vocab to "hidden" representation space and then, after RNN, from hidden representation to target vocab via softmax (or few more layers). This suggest natural symmetry to me. 
Anyway, thanks for your clarifications. An interesting experiment to try.

**Comment by ofirpress** [comment]
> Can't we use constrain (3) to just be transpose of (2) ?
> I haven't really seen that being done, but could be an interesting experiment.

In a [paper](http://aclweb.org/anthology/E/E17/E17-2025.pdf) that we recently presented, we showed that you can tie these matrices, and this improves performance in language modeling.

>It doesn't necessarily contain meaningful representations of words.

We actually show in the paper that the softmax matrix contains representations of words that are better than the ones in the word embedding. 

In addition we showed that in translation models, although this weight tying doesn't help performance wise, it reduces the model's size without harming its performance. 

Lastly, if the source and target vocabularies are very similar (for example when using BPE), than all three of the matrices you mention can be tied (we call this Three Way Weight Tying), and this again massively lowers the amount of parameters while not hurting performance at all. It's already available in Nematus, and I'm sure it would also be useful in this repository.
